[KOGITO-4469] - Conditions should reflect latest evaluated status

vaibhavjainwiz opened the pull request:
Many thanks for submitting your Pull Request :heart:! 

Jira issue : https://issues.redhat.com/browse/KOGITO-4469

Please make sure your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains a link to the JIRA issue
- [x] Pull Request contains a description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster
- [x] You've added a [RELEASE_NOTES.md](RELEASE_NOTES.md) entry regarding this change

sutaakar commented on core/framework/condition_collection.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+func (c *conditionCollection) pushCondition(newCondition api.ConditionInterface) {
+	if len(c.conditions) >= maxBufferCondition {
```
What is the purpose of the bugger here, when every condition type is available just once?

vaibhavjainwiz replied:
`maxBufferLimit` was already there in previous implementation that why I keep it.
yes you are right we can remove it now.

vaibhavjainwiz replied:
done

Kaitou786 commented:
@vaibhavjainwiz If I went though this correctly we will only have the latest `type` in the status. It means if for failure even if the `KogitoRuntime` have multiple failures we will only display the last one(the latest that the reconciler found). So the developer would first fix the latest issue(the only one he can see in the status), and later he would find that there were other failures as well which needs to be fixed. IMO we can keep the same `type/conditions` in the status list if their `message` is different specially in case of failures.

sutaakar commented:
Failure and success conditions have different type, as a result both failure and success conditions are shown when application starts correctly:
```
    - lastTransitionTime: '2021-03-11T09:54:05Z'
      message: >-
        imagestreams.image.openshift.io "process-springboot-example" already
        exists
      reason: ReconciliationFailure
      status: 'False'
      type: Failed
    - lastTransitionTime: '2021-03-11T09:54:06Z'
      status: 'True'
      type: Provisioning
    - lastTransitionTime: '2021-03-11T09:54:58Z'
      status: 'True'
      type: Deployed
```
Shouldn't we have same type for error and successful deployment? That way used doesn't have to investigate timestamps to find out whether the application runs successfully or not.

ricardozanini reviewed (changes requested):
@vaibhavjainwiz this looks OK, but I'd rather move to k8s core API instead with `metav1.Condition`. So we can remove completely our own conditions structure. This must be done for every resource (CRD) we have. Then you can use `metav1.StatusFailure` and ` metav1.ConditionFalse` (and vice-versa) to define the right cause. We still have the `errorFor...` functions to define the correct message and to control the reconciliation logic.

Plus, use `SetConditionsMeta` from `k8s.io/apimachinery/pkg/api/meta` package to control the definition of each condition.

spolti commented on core/framework/condition_collection_test.go:
```diff
@@ -0,0 +1,62 @@
+	"testing"
+)
+
+func TestConditionCollection(t *testing.T) {
```
what is the purpose of this test? shouldn't int be comparing what has been added instead print it?

vaibhavjainwiz replied:
I add those print stmt for my quick testing but I forget to replace them with asserts in final PR. I fixed them now.

spolti commented on core/framework/condition_collection.go:
```diff
@@ -0,0 +1,51 @@
+func (c *conditionCollection) AddCondition(newCondition api.ConditionInterface) []api.ConditionInterface {
+	for i, condition := range c.conditions {
+		if condition != nil && condition.GetType() == newCondition.GetType() {
+			c.removeOldCondition(i)
```
shouldn't be better keep, like, the latest X conditions here?

spolti replied:
maybe set a limit and use FILO?

vaibhavjainwiz replied:
Previous it was implemented in that manner only. But in jira the expectation is to behave like a collection of the latest status of each different type.

ruromero replied:
Each condition should be a snapshot of the latest status. It should be have like a key-value map, not like a history

ricardozanini replied:
Please see: https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go#L30

We should use this behavior and remove the `buffer` thing.

ricardozanini replied:
We have access to this function, it's just a matter of updating our Status conditions.

spolti replied:
I see, ok, is this question resolved?

vaibhavjainwiz replied:
As discussed, I had implemented condition as a snapshot of the latest status and use `k8s.io/apimachinery/pkg/api/meta` Condition object

spolti reviewed (changes requested):
nice work, just a few comments.

vaibhavjainwiz commented:
> `SetConditionsMeta` from `k8s.io/apimachinery/pkg/api/meta`

@ricardozanini I didn't found `SetConditionsMeta` in `k8s.io/apimachinery/pkg/api/meta`. Could you please help me to find it.

vaibhavjainwiz commented:
@ricardozanini I had replaced our custom Condition struct with `metav1.Condition` in all CRDs. There is a catch for KogitoInfra, all crds except kogitoInfra have collection of Conditions inside their status but KogitoInfra only contain single Condition object.

ricardozanini commented:
> @ricardozanini I had replaced our custom Condition struct with `metav1.Condition` in all CRDs. There is a catch for KogitoInfra, all crds except kogitoInfra have collection of Conditions inside their status but KogitoInfra only contain single Condition object.

Yeah, that's wrong. We should have a collection there as well. :)

Can you change it?

ricardozanini commented:
> `SetConditionsMeta`

Sure!

https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go#L30

ricardozanini commented on go.mod:
```diff
@@ -33,24 +33,24 @@ require (
 	knative.dev/eventing v0.18.0
 	knative.dev/pkg v0.0.0-20200922164940-4bf40ad82aab
-	sigs.k8s.io/controller-runtime v0.6.3
+	sigs.k8s.io/controller-runtime v0.6.1
```
Why downgrade here?

vaibhavjainwiz replied:
undo it. To setup complatablity I had made changes in many pkg version. I might miss to reset it

vaibhavjainwiz commented:
> > @ricardozanini I had replaced our custom Condition struct with `metav1.Condition` in all CRDs. There is a catch for KogitoInfra, all crds except kogitoInfra have collection of Conditions inside their status but KogitoInfra only contain single Condition object.
> 
> Yeah, that's wrong. We should have a collection there as well. :)
> 
> Can you change it?

KogitoInfra doesn't provision any k8s object. It just connect with existing resources. Currently we are only maintaing two states(success/failure) on bases if KogitoInfra could be able to connect with provided resource and that enough imformation for user. I not sure if KogitoInfra requires collection of Conditions. Wdyt?

ricardozanini commented on go.mod:
```diff
@@ -33,24 +33,24 @@ require (
-	k8s.io/apimachinery v0.18.8
+	k8s.io/api v0.20.4
+	k8s.io/apiextensions-apiserver v0.20.1
+	k8s.io/apimachinery v0.20.4
```
Hmmm this is tricky.. we might lose compatibility with older kube versions. Please, just make sure that we are safe:
https://github.com/kubernetes/client-go#compatibility-matrix

vaibhavjainwiz replied:
I had tested this functionality on k8s v1.20.0. Do I need to test it with some other k8s version?

ricardozanini replied:
I think we are fine up to 1.18. @sutaakar, the OCP we run our BDD is based on which k8s version?

sutaakar replied:
good question, @Kaitou786 could know that

Kaitou786 replied:
Currently we use OCP 4.6.x for our BDD tests which is based on kubernetes 1.19 as stated in the [docs](https://docs.openshift.com/container-platform/4.6/release_notes/ocp-4-6-release-notes.html#ocp-4-6-about-this-release)

vaibhavjainwiz replied:
I had tested these changes on kubernetes 1.19. They are working fine.

ricardozanini reviewed (commented):
Just small comments. Great work man!

ricardozanini commented:
> > > @ricardozanini I had replaced our custom Condition struct with `metav1.Condition` in all CRDs. There is a catch for KogitoInfra, all crds except kogitoInfra have collection of Conditions inside their status but KogitoInfra only contain single Condition object.
> > 
> > 
> > Yeah, that's wrong. We should have a collection there as well. :)
> > Can you change it?
> 
> KogitoInfra doesn't provision any k8s object. It just connect with existing resources. Currently we are only maintaing two states(success/failure) on bases if KogitoInfra could be able to connect with provided resource and that enough imformation for user. I not sure if KogitoInfra requires collection of Conditions. Wdyt?

I think we might have to implement the collection since it's a "pattern" for Kubernetes resources: https://book-v1.book.kubebuilder.io/basics/status_subresource.html

We need to keep a history of conditions and the last time they transitioned.

vaibhavjainwiz commented on api/conditions.go:
```diff
@@ -49,20 +48,9 @@ const (
-	SetConditions(conditions []ConditionInterface)
+	GetConditions() []metav1.Condition
+	AddCondition(condition metav1.Condition)
+	RemoveCondition(index int)
```
I had removed `AddCondition` and `RemoveCondition` methods but we still requires `SetConditions` method

vaibhavjainwiz commented on core/kogitobuild/status.go:
```diff
@@ -64,72 +45,145 @@ func NewStatusHandler(context *operator.Context) StatusHandler {
 }
 
 func (s *statusHandler) HandleStatusChange(instance api.KogitoBuildInterface, err error) {
```
done

vaibhavjainwiz commented:
> > > > @ricardozanini I had replaced our custom Condition struct with `metav1.Condition` in all CRDs. There is a catch for KogitoInfra, all crds except kogitoInfra have collection of Conditions inside their status but KogitoInfra only contain single Condition object.
> > > 
> > > 
> > > Yeah, that's wrong. We should have a collection there as well. :)
> > > Can you change it?
> > 
> > 
> > KogitoInfra doesn't provision any k8s object. It just connect with existing resources. Currently we are only maintaing two states(success/failure) on bases if KogitoInfra could be able to connect with provided resource and that enough imformation for user. I not sure if KogitoInfra requires collection of Conditions. Wdyt?
> 
> I think we might have to implement the collection since it's a "pattern" for Kubernetes resources: https://book-v1.book.kubebuilder.io/basics/status_subresource.html
> 
> We need to keep a history of conditions and the last time they transitioned.

@ricardozanini I had implemented collection of condition for KogitoInfra. But regarding your comment `keep a history of conditions and the last time they transitioned`, It will not be there.  As describe by @ruromero `Each condition should be a snapshot of the latest status. It should be have like a key-value map, not like a history`. 
For example in KogitoRuntime, now we are maintaining three type of condition `Failed`, `Provisioning` & `Deployed`. On each reconcilation cycle, we found out whats the current state of these 3 condition and set value of all 3 condition every time. So no history details.

sutaakar commented on core/infrastructure/deployment.go:
```diff
@@ -78,3 +79,12 @@ func (d *deploymentHandler) IsDeploymentAvailable(key types.NamespacedName) (boo
+	if err != nil {
+		return 0, err
+	}
+	return deployment.Status.Replicas, nil
```
```suggestion
	return deployment.Status.AvailableReplicas, nil
```

vaibhavjainwiz replied:
done

sutaakar commented on core/kogitoinfra/status_test.go:
```diff
@@ -0,0 +1,57 @@
+	conditions := *instance.Status.Conditions
+	assert.Equal(t, 2, len(conditions))
+	assert.Equal(t, string(api.KogitoInfraFailure), conditions[0].Type)
+	assert.Equal(t, string(api.KogitoInfraSuccess), conditions[1].Type)
```
Please check also status to make sure what are the values of those types.

vaibhavjainwiz replied:
done

sutaakar commented:
The code looks ok, need to test it....

ricardozanini commented:
> @ricardozanini I had implemented collection of condition for KogitoInfra. But regarding your comment `keep a history of conditions and the last time they transitioned`, It will not be there. As describe by @ruromero `Each condition should be a snapshot of the latest status. It should be have like a key-value map, not like a history`.

I think I wasn't clear in my last comment. This pattern is enforced by the `SetConditions` function, whereas we keep the **latest condition** of a given **type**, not all. My bad if I wasn't clear.

See [reference](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties):

> Objects may report multiple conditions, and new types of conditions may be added in the future or by 3rd party controllers. Therefore, conditions are represented using a list/slice of objects, where each condition has a similar structure. This collection should be treated as a map with a key of type.

ricardozanini commented on api/kogitoinfra_types.go:
```diff
@@ -24,10 +24,10 @@ import (
+	// KogitoInfraSuccess ...
+	KogitoInfraSuccess KogitoInfraConditionType = "Success"
+	// KogitoInfraFailure ...
+	KogitoInfraFailure KogitoInfraConditionType = "Failure"
```
We don't need to replicate these constants here. Please use this instead:
https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L719-L722

vaibhavjainwiz replied:
@ricardozanini The link that you had shared above are status but we are using `KogitoInfraSuccess` & `KogitoInfraFailure` as type. In Condition struct we have two fields `type` and `status`. Every type could have its status `success` or `failure`. 
To me as well these type are not looks right, type should be something like `deployed` or `provisioning` which could have their status as `success` or `false`.

ricardozanini replied:
Agreed. We should change this naming then :/

ricardozanini reviewed (approved):
One nitpick, apart from that, looks really good. Many thanks!

sutaakar commented:
A nitpick, I have noticed that `lastTransitionTime` changes in every reconciliation cycle, it should be change only when there is a transition in the condition.

sutaakar commented on api/kogitoservices_types.go:
```diff
@@ -21,6 +21,46 @@ import (
+	SuccessfulDeployedReason KogitoServiceConditionReason = "AtLeastOnePodAvailable"
+	// FailedDeployedReason ...
+	FailedDeployedReason KogitoServiceConditionReason = "NoPodAvailable"
+	// SuccessfulProvisioningReason ...
+	SuccessfulProvisioningReason KogitoServiceConditionReason = "RequestedReplicasNotEqualToAvailableReplicas"
+	// FailedProvisioningReason ...
+	FailedProvisioningReason KogitoServiceConditionReason = "RequestedReplicasEqualToAvailableReplicas"
```
```suggestion
	SuccessfulProvisioningReason KogitoServiceConditionReason = "RequestedReplicasEqualToAvailableReplicas"
	// FailedProvisioningReason ...
	FailedProvisioningReason KogitoServiceConditionReason = "RequestedReplicasNotEqualToAvailableReplicas"
```

sutaakar replied:
IMHO the reason value should be something more abstract. Provisioning can be set to true also in case there is some  error during reconciliation, while pods are already up and running.
Maybe something like:
```suggestion
	// SuccessfulProvisioningReason ...
	SuccessfulProvisioningReason KogitoServiceConditionReason = "ProvisioningInProgress"
	// FinishedProvisioningReason ...
	FinishedProvisioningReason KogitoServiceConditionReason = "RequestedReplicasEqualToAvailableReplicas"
	// FailedProvisioningReason ...
	FailedProvisioningReason KogitoServiceConditionReason = "ProvisioningFailure"
```

sutaakar commented:
@vaibhavjainwiz In case of some error the KogitoInfra shows both success and failure types:
```
    - lastTransitionTime: '2021-03-18T08:38:21Z'
      message: >-
        Infinispan resource(kogito-infinispan3) not found in namespace
        ksuta-test
      reason: ResourceNotFound
      status: 'True'
      type: Failure
    - lastTransitionTime: '2021-03-18T08:38:21Z'
      message: ''
      reason: ResourceNotFound
      status: 'False'
      type: Success
```
Does it have sense to show both? What about just using Success type with possible message in case of an error?

sutaakar commented on core/kogitobuild/status.go:
```diff
@@ -64,114 +44,136 @@ func NewStatusHandler(context *operator.Context) StatusHandler {
+		Type:               string(api.KogitoBuildSuccessful),
+		Status:             status,
+		LastTransitionTime: metav1.Now(),
+		Reason:             string(status),
```
Results in:
```
    - lastTransitionTime: '2021-03-18T08:46:42Z'
      message: ''
      reason: 'False'
      status: 'False'
      type: Successful
```
IMHO we should provide some meaningful reason like BuildInProgress (for the case when build is still in progress), possibly some error reason too for the case of a build error.

sutaakar commented on core/kogitobuild/status.go:
```diff
@@ -64,114 +44,136 @@ func NewStatusHandler(context *operator.Context) StatusHandler {
+		Type:               string(api.KogitoBuildRunning),
+		Status:             status,
+		LastTransitionTime: metav1.Now(),
+		Reason:             string(status),
```
Same here.

sutaakar commented:
Apart from the points above (and the nitpick `lastTransitionTime` being refreshed in every reconciliation) looks good to me.
